### PR TITLE
Quick access to bookmarks and messages on user menu.

### PR DESCRIPTION
**Pull request by nbianca**
This is work in progress. For now, basic support for bookmarks has been implemented. I will continue with messages and then add styling. Any feedback is much appreciated.

### CODE REVIEW

**Comment by tgxworld** [comment]
screenshot from 2017-03-02 15-21-11

Isn't the bookmarks and messages link available from the user-menu already? :thinking:

**Comment by nbianca** [comment]
Hello @tgxworld,

This is actually related to [this](https://meta.discourse.org/t/quick-access-to-bookmarks-and-messages-on-user-menu/32787/6). The desired functionality is something like this:

bookmarks

messages

**Comment by nbianca** [comment]
Hello,

I know that this is still work in progress, but before continuing I would like to hear some feedback. For now, the UI is not really how it should be, but I was interested if the way I implemented the changes is fine with you.

File: app/assets/javascripts/discourse/widgets/user-menu.js.es6 [hunk 1]
```diff
@@ -81,21 +85,149 @@ createWidget('user-menu-links', {
...
+
+  defaultState() {
+    return { loading: false, loaded: false, content: [] };
+  },
+
+  refreshBookmarks(state) {
+    if (state.loading) { return; }
+
+    const { currentUser } = this;
+    let findUrl = Discourse.getURL('/user_actions.json?username=' + currentUser.username + "&filter=" + Discourse.UserAction.TYPES.bookmarks);
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/widgets/user-menu.js.es6]: You can just pass the path into `ajax` which [will call getURL](https://github.com/discourse/discourse/blob/master/app/assets/javascripts/discourse/lib/ajax.js.es6#L103) for you

Comment (tgxworld) [inline app/assets/javascripts/discourse/widgets/user-menu.js.es6]: Also params can be passed in as `data` to the `ajax` function

File: app/assets/javascripts/discourse/widgets/user-menu.js.es6 [hunk 2]
```diff
@@ -81,21 +85,149 @@ createWidget('user-menu-links', {
...
+    if (!state.loaded) {
+      this.refreshBookmarks(state);
+    }
+
+    if (state.loading) {
+      return h('div.spinner-container', h('div.spinner'));
+    }
+
+    return h('ul.menu-bookmarks', [
+      state.content.map(b => h('li', [
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/widgets/user-menu.js.es6]: I'm not quite sure what the variable `b` represents here. Usually we want to name variables such that it is obvious on first look what the variable is holding.

File: app/assets/javascripts/discourse/widgets/user-menu.js.es6 [hunk 3]
```diff
@@ -81,21 +85,149 @@ createWidget('user-menu-links', {
...
+        h('i', { className: "fa fa-bookmark" }),
+        h('a.bookmark', { attributes: { href: b.get('postUrl') } }, b.title)
+      ]))
+    ]);
+  },
+
+});
+
+
+export default createWidget('user-private-messages', {
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/widgets/user-menu.js.es6]: Quite abit of code is being duplicated between the two widgets. I would look at ways to try and reduce the duplication in the code.

File: app/assets/javascripts/discourse/widgets/user-menu.js.es6 [hunk 4]
```diff
@@ -81,21 +85,149 @@ createWidget('user-menu-links', {
...
+    const panels = [];
+    panels.push(this.attach('user-menu-links', { path }));
+
+
+    if (this.state.userBookmarks) {
+      panels.push(this.attach('user-bookmarks', { path }));
+    } else if (this.state.privateMessages) {
+      panels.push(this.attach('user-private-messages', { path }));
+    } else {
+      panels.push(this.attach('user-notifications', { path }));
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/widgets/user-menu.js.es6]: UI wise, how do we toggle back to user's notifications?

File: app/assets/javascripts/discourse/widgets/user-menu.js.es6 (hunk 3 above)
Comment (nbianca) [inline app/assets/javascripts/discourse/widgets/user-menu.js.es6]: Yeah, I noticed that the structure of those two is similar, but if I were to create an abstract widget and then extend it, the code would get much more messier.

File: app/assets/javascripts/discourse/widgets/user-menu.js.es6 (hunk 4 above)
Comment (nbianca) [inline app/assets/javascripts/discourse/widgets/user-menu.js.es6]: You click again on bookmark or private message icon to toggle.

**Comment by nbianca** [comment]
I did some changes to the UI.

**Comment by tgxworld** [comment]
Hi @nbianca, is this still WIP?

**Comment by nbianca** [comment]
Hi @tgxworld,

Yes, sorry. This is still WIP. However, it should be ready soon.

**Comment by nbianca** [comment]
Hello @tgxworld,

I made more changes as per Jeff's review. I think it is done.

**Comment by tgxworld** [comment]
Hi @nbianca, I think some stuff might have broke since the last update.

screenshot from 2017-04-06 15-27-26

screenshot from 2017-04-06 15-27-53

**Comment by nbianca** [comment]
Sorry, this got fixed.

**Comment by coding-horror** [comment]
Anything else to do here?

**Comment by tgxworld** [comment]
@nbianca I'm still seeing the same issue

screenshot from 2017-07-17 15-47-17

**Comment by nbianca** [comment]
Hi. I think some refactoring was done there and something broke (or maybe it was not implemented properly from the beginning). I will have to redo some of the changes.

**Comment by nbianca** [comment]
Hello. This should be fixed now.

**Comment by tgxworld** [comment]
Hi @nbianca, I would suggest adding some JavaScript tests for this feature :) I pulled down the feature locally again and the user-menu stopped working with a generic error in the console.

screenshot from 2017-07-21 09-42-04

**Comment by SamSaffron** [comment]
we definitely need a few tests here to finish this off.

**Comment by SamSaffron** [comment]
This is getting there but still needs some work.

image

- I am seeing "unread" coloring on my bookmarks that does not make sense. The lines should be white

- Spacing feels too tight

- I would always have a link to the bookmarks page even if you only have 2 bookmars

Same goes for 

image


Need more spacing. Need a link to PMs as well (cause there are other cases you want to go to pms eg: pms to groups or compose PM) Also all PMs are read. 

I am also not seeing any highlighting on the selected "tab" so we can not easily tell what is selected. 

Another minor issue is that the only way to get back to the list of notifications is to click the selected tab again, something that is not plainly obvious.

**Comment by ZogStriP** [comment]
@nbianca would you mind posting a couple screenshots of the latest version?

**Comment by nbianca** [comment]
This is how it looks now. The icons are blue when they are selected. On hover, they have that yellow background as links usually have.

screen shot 2017-09-04 at 18 32 29
screen shot 2017-09-04 at 18 30 47
screen shot 2017-09-04 at 18 30 57

**Comment by mcwumbly** [comment]
This is cool!

As part of this, or as a follow-up, I think it might be nice to move the
settings icon to be left-aligned next to the profile link.

That way the icons that are quick access are on one side, apart from the
ones that navigate immediately to another page.

**Comment by tgxworld** [comment]
I just reviewed this code, the logic used to fetch the bookmarks and messages need to be similar to that of the notifications. Otherwise, we're not caching any of the existing results and will always be sending a request to the server even if there are no new bookmarks/messages.

**Comment by coding-horror** [comment]
Any ETA on this, it has been open a while, are we blocking?

**Comment by erlend-sh** [comment]
I think Bianca is occupied with studies at the moment.

**Comment by SamSaffron** [comment]
I think we have a fair bit of work left on this, closing it for now and if @nbianca frees up after studies we can look at revisiting

**Pull request closed without merging**